test-app-native-modules.gradle

```diff
@@ -0,0 +1,271 @@
+    if (this.reactNativeModules != null) return this.reactNativeModules
+
+    ArrayList<HashMap<String, String>> reactNativeModules = new ArrayList<HashMap<String, String>>()
+
+    /**
+     * Resolve the CLI location from Gradle file
+     *
+     * @todo: Sometimes Gradle can be called outside of the JavaScript hierarchy (-p flag) which
+     * will fail to resolve the script and the dependencies. We should resolve this soon.
+     *
+     * @todo: `fastlane` has been reported to not work too.
+     */
+    def cliResolveScript = "console.log(require('react-native/cli').bin);"
+    String[] nodeCommand = ["node", "-e", cliResolveScript]
+    def cliPath = this.getCommandOutput(nodeCommand)
+
+    String[] reactNativeConfigCommand = ["node", cliPath, "config"]
+    def reactNativeConfigOutput = this.getCommandOutput(reactNativeConfigCommand)
+
+    def json
+    try {
+      json = new JsonSlurper().parseText(reactNativeConfigOutput)
+    } catch (Exception exception) {
+      throw new Exception("Calling `${reactNativeConfigCommand}` finished with an exception. Error message: ${exception.toString()}. Output: ${reactNativeConfigOutput}");
+    }    
+
+    json["dependencies"].each { name, value ->
+      def platformsConfig = value["platforms"];
+      def androidConfig = platformsConfig["android"]
+
+      if (androidConfig != null && androidConfig["sourceDir"] != null) {
+        this.logger.info("${LOG_PREFIX}Automatically adding native module '${name}'")
+
+        HashMap reactNativeModuleConfig = new HashMap<String, String>()
+        reactNativeModuleConfig.put("name", name)
+        reactNativeModuleConfig.put("nameCleansed", name.replaceAll('[~*!\'()]+', '_').replaceAll('^@([\\w-.]+)/', '$1_'))
+        reactNativeModuleConfig.put("androidSourceDir", androidConfig["sourceDir"])
+        reactNativeModuleConfig.put("packageInstance", androidConfig["packageInstance"])
+        reactNativeModuleConfig.put("packageImportPath", androidConfig["packageImportPath"])
+        this.logger.trace("${LOG_PREFIX}'${name}': ${reactNativeModuleConfig.toMapString()}")
+
+        reactNativeModules.add(reactNativeModuleConfig)
+      } else {
+        this.logger.info("${LOG_PREFIX}Skipping native module '${name}'")
+      }
+    }
+
+    return reactNativeModules
+  }
+}
+
+/** -----------------------
+ *    Exported Extensions
+ * ------------------------ */
+
+def autoModules = new ReactNativeModules(logger, jsAppDir)
+
+ext.applyNativeModulesSettingsGradle = { DefaultSettings defaultSettings, String root = null ->
+  if (root != null) {
+    logger.warn("${ReactNativeModules.LOG_PREFIX}Passing custom root is deprecated. CLI detects root automatically now.");
+    logger.warn("${ReactNativeModules.LOG_PREFIX}Please remove second argument to `applyNativeModulesSettingsGradle`.");
+  }
+  autoModules.addReactNativeModuleProjects(defaultSettings)
+}
+
+ext.applyNativeModulesAppBuildGradle = { Project project, String packageName, String root = null ->
```

Comment:
Upstream the change to the autolinking plugin to specify a package name in parameters. This way we can force the plugin to generate code for a module other than the one that `react-native config` was executed for. An alternative solution is to derive the `packageName` from the `project` that the plugin gets applied to.

Reply:
https://github.com/microsoft/react-native-test-app/projects/1#card-35147467